arangod/IResearch/IResearchAnalyzerFeature.cpp

```diff
@@ -458,22 +471,87 @@ arangodb::aql::AqlValue aqlFnTokens(arangodb::aql::ExpressionContext* expression
   auto buffer = irs::memory::make_unique<arangodb::velocypack::Buffer<uint8_t>>();
 
   if (!buffer) {
-    irs::string_ref const message = "failure to allocate result buffer while "
-                                    "computing result for function 'TOKENS'";
+    irs::string_ref const message =
+        "failure to allocate result buffer while "
+        "computing result for function 'TOKENS'";
 
     LOG_TOPIC("97cd0", WARN, arangodb::iresearch::TOPIC) << message;
     THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_OUT_OF_MEMORY, message);
   }
 
   arangodb::velocypack::Builder builder(*buffer);
-
   builder.openArray();
 
-  while (analyzer->next()) {
-    auto value = irs::ref_cast<char>(values->value());
-
-    arangodb::iresearch::addStringRef(builder, value);
-  }
+  std::deque<arangodb::velocypack::ArrayIterator> arrayIteratorStack;
+  auto current = args[0].slice();
+  do {
+    // stack opening non-empty arrays
+    while (current.isArray() && !current.isEmptyArray()) {
+      arrayIteratorStack.emplace_back(current);
+      builder.openArray();
+      current = arrayIteratorStack.back().value();
+    }
+    // process current item
+    if (current.isString()) {
+      auto const& data = arangodb::iresearch::getStringRef(current);
+      if (!string_analyzer->reset(data)) {
+        auto const message = "failure to reset arangosearch analyzer: ' "s +
+                             static_cast<std::string>(name) +
+                             "' while computing result for function 'TOKENS'";
+
+        LOG_TOPIC("45a2d", WARN, arangodb::iresearch::TOPIC) << message;
+        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, message);
+      }
+      while (string_analyzer->next()) {
+        auto value = irs::ref_cast<char>(string_terms->value());
+        arangodb::iresearch::addStringRef(builder, value);
+      }
+    } else if (current.isBool()) {
+      arangodb::iresearch::addStringRef(
+          builder, arangodb::basics::StringUtils::encodeBase64(irs::ref_cast<char>(
+                       irs::boolean_token_stream::value(current.getBoolean()))));
```

Comment:
I have no idea what this should do. At the very least I've got to see the documentation first.

Reply:
This will give us  internal analyzer representation for given value, just like with strings